### DISCUSSION TITLE
Order of X and y arguments in fit methods

### DISCUSSION THREAD

**Discussion by ngupta23** [alan-turing-institute/sktime#683] [General]
**Is your feature request related to a problem? Please describe.**

Most of sklearn compatible `fit` methods are called using X, y order

```python
# sklearn and (most) sklearn compatible libraries such as imblearn - fit definition
def fit(Xt, yt, **fit_params):
    ...

# call
self._final_estimator.fit(Xt, yt, **fit_params)
```

If the `_final_estimator` is a sktime model, this causes an issue since the X and y arguments get reversed because sktime `fit` methods are called using y, X argument order (reverse of sklearn and sklearn compatible libraries).

```python
# sktime fit methods
def fit(self, y, X=None, fh=None, **fit_params)
```

**Describe the solution you'd like**
It would be great if sktime argument order could be made consistent with sklearn and other sklearn compatible libraries by reversing the order of y and X arguments in fit.

y could default to None and an internal check could be added in the base class to make sure that this to be present, else flag an error.

```python
# sktime fit methods (proposed)
def fit(self, X=None, y=None, fh=None, **fit_params):
    if y is None:
        raise ValueError("y value was not provided")
```

**Describe alternatives you've considered**
I can not think of any.

**Additional context**
Calling sktime models using `imblearn` pipeline which uses sklearn compatible calls [using X, y format](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/4162d2d8ef127a81238ee72ebabac55642daee7c/imblearn/pipeline.py#L281). This causes issues since sktime expects y to be the first argument, so X and y get flipped internally.

**Comment by fkiraly** [comment]
Quick comment: this suggests a major API change!
Reversing `y` and `X` was a conscious API choice due to the differences between tabular learning tasks and forecasting.

Note that in forecasting, `X` is optional while `y` is not, whereas in `sklearn` tabular learning, `y` is optional (in unsupervised instances) while `X` is not!

The rule "optional arguments come last" implies the ordering already.
That is not superseded by "we could make `y` optional too", since it does not change the "moral story", i.e., the semantic optionality.

**Comment by fkiraly** [comment]
pinging @mloning for the forecasting module

**Comment by fkiraly** [comment]
An alternative we discussed internally is whether in forecasting, the symbols `y` and `X` should be exchanged while leaving the semantic meaning intact, i.e., `y`/`Y` for exogenous variables and `X` for endogenous ones. We decided against that due to the common association of the symbol `y` with "target" and `X` with "side information" being stronger than the association of `X` with "core/non-optional data".

**Comment by ngupta23** [comment]
Hi @fkiraly, Thanks for the quick feedback. I understand this was a conscious choice, but making sktime compatible with sklearn and other sklearn compatible API calls would open up more functionality to be built around sktime.

**Comment by mloning** [comment]
scikit-learn's API is for cross-sectional learning, forecasting and other time series tasks are considerably different to cross-sectional learning. I think it makes sense to have a different interface for different tasks. For more general design principles behind sktime, see our [paper](https://arxiv.org/abs/2101.04938). The rationale for the forecasting API is also discussed in more details in this [enhancement proposal](https://github.com/sktime/enhancement-proposals/blob/master/steps/01_forecasting_api/01_forecasting_api.md). Feedback is very welcome! 

Making estimators fully compatible with the scikit-learn interface also requires changing the `predict` method by the way, so you could do something like `fbprophet` does, but I don't think it's very intuitive, and it uses the same symbols, but changes the meaning of the symbols. In scikit-learn, `X` is cross-sectional data whereas in sktime, it'd be time series data. One also has to be careful to apply functionality from the cross-sectional setting to forecasting, because some assumptions do not hold (e.g. rows do not represent i.i.d. samples). 

sktime is still interoperable with scikit-learn through the reduction interfaces (e.g. `ReducedRegressionForecaster`). Why doesn't that work in the above example? What other functionality do you have in mind that doesn't work with the current interface?

**Reply by ngupta23** [reply]
@mloning, We are actually calling the `sktime` `ARIMA` model through a `imblearn` pipeline (so it is not using the native sklearn models). We are also using sklearn's native [`TimeSeriesSplit`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) so there are no assumptions as far as i.i.d. is concerned. 

The reason is that we are trying to wrap sktime models within `PyCaret` library and need to be compliant with sklearn since PyCaret is natively compliant with it.

**Reply by ngupta23** [reply]
Also, just to clarify, I am not implying changing the meaning of X or y in sktime. I am only requesting the reversing of the order of arguments. All internals can remain same, just the order would be changed in the method signature.

**Reply by mloning** [reply]
Out of curiosity, how would you handle `predict` calls?

**Reply by fkiraly** [reply]
@ngupta23, I would strongly discourage wrapping `sktime` estimators under `pycaret` classifier/regressor interface, which is what it sounds like you are trying - because it's a different learning task! Similar how classifiers are different from regressors per interface in `pycaret`.

What would make more sense i.m.o. is to have a wrapper interface *by task*, and that you can make it similar to `sktime` if you want.

**Reply by ngupta23** [reply]
OK, I see the issue now with using this with sklearn methods. It does not like `X = None`. Is this why you wrote your own [`_fit_and_score` function](https://github.com/alan-turing-institute/sktime/blob/0e79faaf9f94cde4a513c99e0000a3ce8df2f8b6/sktime/forecasting/model_selection/_tune.py#L95)?

**Reply by fkiraly** [reply]
Yes, exactly - there's no `X` in `predict`.

**Comment by fkiraly** [comment]
Looks like the discussion above is settled?
I think it's unlikely we'll change the signature soon, of course discussion is always welcome.
